test/project_types/extension/commands/serve_test.rb

```diff
@@ -16,24 +16,60 @@ def setup
 
       def test_defers_serving_to_the_specification_handler
         serve = ::Extension::Commands::Serve.new(@context)
+        stub_specification_handler_options(serve)
         serve.specification_handler.expects(:serve)
         serve.call([], "serve")
       end
 
-      def test_error_raised_if_no_available_ports_found
+      def test_error_raised_if_specification_handler_supports_choosing_port_but_no_ports_available
         serve = ::Extension::Commands::Serve.new(@context)
 
         Tasks::ChooseNextAvailablePort.expects(:call)
           .with(from: ::Extension::Commands::Serve::DEFAULT_PORT)
           .returns(ShopifyCli::Result.failure(ArgumentError))
           .once
+        serve.specification_handler.expects(:choose_port?).returns(true).once
 
         error = assert_raises ShopifyCli::Abort do
           serve.call([], "serve")
         end
 
         assert_includes error.message, @context.message("serve.no_available_ports_found")
       end
+
+      def test_port_not_chosen_if_specification_handler_does_not_support_chosen_port
+        serve = ::Extension::Commands::Serve.new(@context)
+        stub_specification_handler_options(serve)
+        Tasks::ChooseNextAvailablePort.expects(:call).never
+        serve.specification_handler.expects(:serve).once
+        serve.call([], "serve")
+      end
+
+      def test_new_tunnel_started_if_specification_handler_supports_establish_tunnel
+        serve = ::Extension::Commands::Serve.new(@context)
+        stub_specification_handler_options(serve, choose_port: true, establish_tunnel: true)
+        ShopifyCli::Tunnel.expects(:start)
+          .with(@context, port: Extension::Commands::Serve::DEFAULT_PORT)
+          .returns("ngrok.example.com")
+          .once
+        serve.specification_handler.expects(:serve).once
+        serve.call([], "serve")
+      end
+
+      def test_tunnel_not_started_if_specification_handler_does_not_support_establish_tunnel
+        serve = ::Extension::Commands::Serve.new(@context)
+        stub_specification_handler_options(serve)
+        ShopifyCli::Tunnel.expects(:start).never
+        serve.specification_handler.expects(:serve).once
+        serve.call([], "serve")
+      end
+
+      private
+
+      def stub_specification_handler_options(serve, choose_port: false, establish_tunnel: false)
```

Comment:
Totally a matter of personal preference, but I would avoid introducing this helper - Personally, I think it's more important for testing methods to be as self-contained and explicit, even if it requires some repetition.

Reply:
TL;DR I would keep the helper.

Generally, I agree with @joeyfreund's sentiment, but if the helper name clearly states its purpose and abstracts some plumbing away, I'm generally ok with using a helper instead of repeating the same code. This is especially true for helpers that take care of mocking, something that is prone to be brittle and often has to be changed when the internals of our codebase change. Being able to make these change is one place helps to speed up this process.

However, as a rule of thumb, I 💯 percent echo what Joey said. Tests should be as explicit and self-contained as possible.